src/openforms/forms/admin/form.py

```diff
@@ -334,10 +334,38 @@ def remove_from_maintenance_mode(self, request, queryset):
     )
 
     def delete_model(self, request, form):
-        # override for soft-delete
-        form._is_deleted = True
-        form.save(update_fields=["_is_deleted"])
+        """
+        Check if we need to soft or hard delete.
+        """
+        if not form._is_deleted:
+            # override for soft-delete
+            form._is_deleted = True
+            form.save(update_fields=["_is_deleted"])
+        else:
+            fds = list(
+                FormDefinition.objects.filter(
+                    formstep__form=form, is_reusable=False
+                ).values_list("id", flat=True)
+            )
+            form.delete()
+            FormDefinition.objects.filter(id__in=fds).delete()
 
     def delete_queryset(self, request, queryset):
-        # override for soft-delete
-        queryset.update(_is_deleted=True)
+        """
+        Split between soft and hard deletes here.
+
+        The admin has mutually exclusive filters, but let's not rely on that assumption.
+        """
+
+        # soft-deletes
+        queryset.filter(_is_deleted=False).update(_is_deleted=True)
+
+        # hard deletes - ensure we cascade delete the single-use form definitions as well
+        soft_deleted = queryset.filter(_is_deleted=True)
```

Comment:
Please switch soft and hard blocks around to prevent accidental soft->hard deletes in one go.

Reply:
test added and patched